Final/TCPServer.cpp

```diff
@@ -0,0 +1,113 @@
+        error("ERROR binding to socket");
+
+    // Step 3: Listen
+    if (listen(sock_fd, BACKLOG_SIZE) < 0) // The socket will actively listen for incoming connections and queue them up
+        error("ERROR listener failed");
+
+    while (true)
+    {
+        // Reset read_fds
+        fd_set read_fds;
+        FD_ZERO(&read_fds);
+        FD_SET(sock_fd, &read_fds);
+        int maxfd = 0;
+        for (node *open_conn_fd = open_conns->begin(); open_conn_fd != open_conns->end(); open_conn_fd = open_conn_fd->next)
+        {
+            int data = open_conn_fd->data;
+            FD_SET(data, &read_fds); // populate read_fds from open_conns
+            if (data > maxfd) // calculate maxfd
+                maxfd = data;
+        }
+
+        if (sock_fd > maxfd)
+            maxfd = sock_fd;
+
+        std::cout << "Attempting to Select . . ." << std::endl;
+        if (select(maxfd + 1, &read_fds, NULL, NULL, NULL) < 0) // sleeps until there is an available connection and then determines if it must be accepted or read from
+            error("My error message");
+        std::cout << "Connection Selected" << std::endl;
+
+        if (FD_ISSET(sock_fd, &read_fds))
+            accept_connection(sock_fd, open_conns);
+        else
+            handle_connection(read_fds, open_conns, handler);
+    }
+
+    // Close the connection
+    std::cout << "Closing the connection" << std::endl;
+    close(sock_fd);
+}
+
+//Pre: sock_fd is the socket bound to this server
+//Post: accept connection and add file descriptor of connected client to open_conns
+void TCPServer::accept_connection(int sock_fd, LinkedList *open_conns)
+{
+    // Step 4: Accecpt
+    struct sockaddr_in remote_host; // a pointer to a sockaddr struct that will be filled in with the address information of the remote host
+    socklen_t remote_len = sizeof(remote_host); // the size of the sockaddr struct in bytes
+    std::cout << "Attempting to Accept . . ." << std::endl;
+    int remotefd = accept(sock_fd, (struct sockaddr *) &remote_host, &remote_len); // dequeue from sock_fd and accept connection through remotefd
+    if (remotefd < 0)
+        error("ERROR accepting request");
+    std::cout << "Connection Accecpted" << std::endl;
+    open_conns->push(remotefd);
+}
+
+//Pre: open_conns contains fd that is also in with read_fds
+//Post: determine which fd has been updated and echo data
+void TCPServer::handle_connection(fd_set read_fds, LinkedList *open_conns, std::function<bool (int)> handler)
+{
+    int fd;
+    for (node *open_conn_fd = open_conns->begin(); open_conn_fd != open_conns->end(); open_conn_fd = open_conn_fd->next)  // iterate through fdlist
+    {
+        fd = open_conn_fd->data;
+        if (FD_ISSET(fd, &read_fds))   // check if file descriptor is ready to be read from
+            break;
+    }
+    std::cout << "Handeling Received Data" << std::endl;
+
+    thread_pool->enqueue([fd, handler, open_conns]
+    {
+        if(handler(fd))
```

Comment:
hmmm. this may run into a race condition. if we don't remove the fd from open_conns before entering the thread, then we will listen to this fd again in the next select statement. if lots of data is being sent from the client to this fd, select will return again saying this fd is ready, and we may end up spawning two threads that read from this same fd. 

to avoid this, we need to preemptively remove the fd from open_conns before spawning off the thread. and then inside the thread, if the handler returns true, we need to call `close` on the fd, otherwise if the handler returns false, we need to add back the fd to open_conns.

note that doing the approach above will cause us to run into yet another race condition lol. this race condition involves manipulating open_conns from multiple threads. if the main thread along with any other thread we spawn are all adding and removing things from open_conns at the same time, it's possible that we get into race conditions that cause data to be lost or corrupted in the linked list. this is because our current implementation of linked list is **not thread safe** (https://en.wikipedia.org/wiki/Thread_safety). to fix this, we can either synchronize access to the linked list in the TCPServer (using locks), or make the linked list a thread safe data structure (again using locks). i prefer the latter approach since it's a bit cleaner.

now to make the linkedlist threadsafe, we can either build in thread safety into the linkedlist (which will slow it down), or keep our linkedlist implementation pure and expose a new SyncLinkedList (short for synchronous linked list) that builds on top of our linked list implementation using [composition](https://en.wikipedia.org/wiki/Object_composition). Note that while OOP inheritence was popular in the past, composition is the preferred way of programming in modern days (https://en.wikipedia.org/wiki/Composition_over_inheritance) as it's much more scalable.

therefore, let's create a new SyncLinkedList:

```
#ifndef SYNC_LINKED_LIST_H
#define SYNC_LINKED_LIST_H

class LinkedList;
struct node;

class SyncLinkedList {
public:
    // Constructor
    SyncLinkedList();
    // Destructor
    ~SyncLinkedList();

    // Pushes node to front of list
    void push(int data);
    // Removes the first instance of a node with the given data
    void remove(int data);

    // Exposed to allow for easy iteration through the linked list
    // If the iter handler returns false, the iteration will stop.
    void for_each(std::function<bool (node *)> iter);
private:
    LinkedList *linked_list;
    std:: shared_mutex rwm; 
};

#endif
```

The SyncLinkedList should just be a light wrapper around the operations of the underlying linked list, but that instead just locks on operations to ensure that only one thread has access. For many data structures, we want to allow behavior such that only a single thread has access to the data structure during any write operations (since they are mutating the underlying data), but otherwise, multiple threads can read from the data structure at any point in time. Take a look at these three data structures (along with google) to see if you can figure out how to implement this type of behavior (https://en.cppreference.com/w/cpp/thread/shared_mutex, ,https://en.cppreference.com/w/cpp/thread/shared_lock, https://en.cppreference.com/w/cpp/thread/lock_guard). Note that because we will need the ability to lock reads if a write is happening, the SyncLinkedList doesn't expose a `begin` and `end` method that allows for free iteration, but instead now exposes a `for_each` method that will call the passed in `iter` function on each iteration, allowing us to fully control how the linked list handles iteration.